db/hints/resource_manager.cc

```diff
@@ -186,19 +192,38 @@ future<> resource_manager::start(shared_ptr<service::storage_proxy> proxy_ptr, s
 }
 
 void resource_manager::allow_replaying() noexcept {
+    set_replay_allowed();
     boost::for_each(_shard_managers, [] (manager& m) { m.allow_replaying(); });
 }
 
 future<> resource_manager::stop() noexcept {
+    unset_running();
     return parallel_for_each(_shard_managers, [](manager& m) {
         return m.stop();
     }).finally([this]() {
         return _space_watchdog.stop();
     });
 }
 
-void resource_manager::register_manager(manager& m) {
-    _shard_managers.insert(m);
+future<> resource_manager::register_manager(manager& m) {
+    const auto [it, inserted] = _shard_managers.insert(m);
+    if (!inserted) {
+        // Already registered
+        return make_ready_future<>();
+    }
+    if (!running()) {
+        return make_ready_future<>();
+    }
+
+    return m.start(_proxy_ptr, _gossiper_ptr, _ss_ptr).then([this, &m] {
+        return with_semaphore(_operation_lock, 1, [this, &m] () {
+            return prepare_per_device_limits().then([this, &m] {
```

Comment:
> I'm not sure I understand this code: all managers are started from resource_manager::start().
> The same is about calling prepare_per_device_limits().
> And the call sequence in storage_proxy::start_hints_manager(..) is that we first call resource_manager::register_manager(...) and then we call resource_manager::start().

The idea is that now, `resource_manager::register_manager(...)` can also be called _after_ the `resource_manager` is started. This will happen if you didn't enable regular hints at boot time, but later tried to do this via HTTP API.

I adjusted the `resource_manager::register_manager(...)` method so that registering of a hints manager is possible after boot. It also automatically starts and allows hints replay, if needed - we need the `resource_manager::state` enum because of that.

It looks like Avi merged this PR already, so I'll send a followup which add comments with explanation.

> prepare_per_device_limits() doesn't seem to be idempotent, nor is manager::start() (it's actually a bug to call it twice).

You are right about `prepare_per_device_limits()`. I mistakenly thought it was idempotent because it checks if manager's device_id is already present in the `_per_device_limits_map`. I didn't notice that managers are unconditionally appended to the `per_device_limits::managers` field.

The impact of this seems to be that the `resource_manager` will count the space used by view hints segments twice, if we enable the regular hints manager in runtime. I'll fix this in the followup, too.

The `hints::manager` shouldn't be started twice. `resource_manager::register_manager(...)` does nothing if it sees that the hints manager you are trying to register was registered before.